### PR TITLE
internal: Bring back JodChild into flychecking for cancellation

### PR DESCRIPTION
**Pull request by Veykril**
cc https://github.com/rust-lang/rust-analyzer/pull/10517/files#r895241975

### CODE REVIEW

File: crates/flycheck/src/lib.rs [hunk 1]
```diff
@@ -153,14 +157,24 @@ impl FlycheckActor {
         while let Some(event) = self.next_event(&inbox) {
             match event {
                 Event::Restart(Restart) => {
+                    // Drop and cancel the previously spawned process
+                    self.cargo_handle.take();
```
Comment (matklad) [inline crates/flycheck/src/lib.rs:161]:
```suggestion
                    
                    if let Some(handle) = self.cargo_handle.take() {
                        // Drop and cancel the previously spawned process
                        handle.kill();
                        let _ = handle.wait();
                    }
```

Comment's great, but I bet explicit code is event better!

File: crates/flycheck/src/lib.rs [hunk 2]
```diff
@@ -153,14 +157,24 @@ impl FlycheckActor {
...
                     self.cancel_check_process();
 
                     let command = self.check_command();
-                    tracing::info!("restart flycheck {:?}", command);
-                    self.cargo_handle = Some(CargoHandle::spawn(command));
-                    self.progress(Progress::DidStart);
+                    let command_f = format!("restart flycheck {command:?}");
+                    match CargoHandle::spawn(command) {
+                        Ok(cargo_handle) => {
+                            tracing::info!("{}", command_f);
```
Comment (matklad) [inline crates/flycheck/src/lib.rs:170]: Note, `do X` in logs is ambiguous: did the log line happen before, or after `X`?

I quite like VSCode naming here: 

```
will restart flycheck
did  restart flycheck
```

File: crates/stdx/src/lib.rs [hunk 3]
```diff
@@ -157,6 +159,16 @@ impl Drop for JodChild {
 }
 
 impl JodChild {
+    pub fn spawn(mut command: Command) -> std::io::Result<Self> {
```
Comment (matklad) [inline crates/stdx/src/lib.rs:162]:
```suggestion
    pub fn spawn(mut command: Command) -> sio::Result<Self> {
```

File: crates/stdx/src/lib.rs [hunk 4]
```diff
@@ -157,6 +159,16 @@ impl Drop for JodChild {
...
+        command.spawn().map(Self)
+    }
+
+    pub fn wait(self) -> std::io::Result<std::process::ExitStatus> {
+        let mut inner = self.into_inner();
+        let _ = inner.kill();
+        inner.wait()
+    }
```
Comment (matklad) [inline crates/stdx/src/lib.rs:170]: probably best to separate kill&wait

**Review by matklad — COMMENTED** [review]
LGTM, but this wants some manual check tht the behavior is OK (that, as soon as you start typing, in-process check is killed, and that is not restarted until you stop typing)

**Comment by Veykril** [comment]
Any tips on how I would go about testing the behaviour (in windows to be specific), I haven't fiddled too much with process inspection.

**Comment by Veykril** [comment]
Typing doesn't have any effect here, but repeatedly saving a file does kill the process if it is outstanding until no more saving occurs.

**Pull request merged**